### ISSUE TITLE
Inference time differs much with same render_factor while device.set(device=DeviceId.GPU0)

### ISSUE THREAD

**Issue by hiterjoshua** [jantic/DeOldify#257]
Awesome work! Still, during my inference time with the function **save_transformed_image**, its really weird that if I add or delete some sentence copied from ColorizeTrainingStable.ipynb as below, the inference time would differs from 10sec/pic to less than 1sec/pic with the same render_factor set as 25.
```
#NOTE:  This must be the first call in order to work properly!
from deoldify import device
from deoldify.device_id import DeviceId
#choices:  CPU, GPU0...GPU7
device.set(device=DeviceId.GPU0)
```
So I am wandering if you noticed this phenomenon, and could you share me with some analysis of this phenomenon? Cause I read the code and noticed in _device.py, you mentioned `This must be called first before any torch imports in order to work properly!`, but I still cannot figure out why the inference time declined sharply.

Thanks a lot!

**Comment by hiterjoshua** [comment]
@jantic could you please share me with your strategy here? cause I find the old test file committed last year does not have the codes mentioned above. and it did decrease the inference time a lot.

 And also, any advice from any person is highly appreciated and welcomed!!!

**Comment by jantic** [comment]
@hiterjoshua  I'm not clear on what your confusion is on exactly.  That code your citing above makes the difference between running inference on GPU and running it on CPU (which is default if those lines are omitted). So that would explain why it's running faster when you add the lines.

This device code was added back in March:  [code](https://github.com/jantic/DeOldify/blob/master/deoldify/_device.py)

The reason why I have it defaulting to CPU if the GPU is not specified is because that's the common denominator on all machines.  It'll at least run, albeit very slowly.

Basically what I was trying to achieve here was to make a reliable way for people to switch between different GPUs or CPU with a single line of code.  Previously the code that needed to be changed was a bit scattered and counterintuitive.  For example, behind the scenes you specific CPU like this (and this is what you used to have to do manually):

`os.environ['CUDA_VISIBLE_DEVICES']=''`

I'd rather have it be more explicit so now users can just do this at the top:

`device.set(device=DeviceId.CPU)`

I hope that makes sense.

**Comment by hiterjoshua** [comment]
@jantic Thanks for your kind reply, perhaps my explanation is not clear and what happened to me is slightly different from your reply above.

To make it clear, the code I use:

```
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'


from deoldify.visualize import *
plt.style.use('dark_background')
torch.backends.cudnn.benchmark=True

# Weights are implicitly read from ./models/ folder
weights_name_nopth = 'imagenet-models/StableModel_imagenet_gen_0'
weights_name = os.path.join(os.path.dirname(os.path.abspath(__file__)), weights_name_nopth)
result_dir = 'output'

colorizer = ModelImageVisualizer(weights_name, result_dir)

##3NOTE:  Max is 45 with 11GB video cards. 35 is a good default
render_factor=25
##3NOTE:  Make source_url None to just read from file at ./video/source/[file_name] directly without modification

source_path = 'test-rgb'

result_path = None
for image_path in os.listdir(lab_path):
    image_path = lab_path + '/' + image_path

    import time
    t0 = time.time()
    result_path = colorizer.save_transformed_image(path=image_path, render_factor=render_factor, compare=True)

    t1 = time.time()
    print(" Timer: %.4f sec ." % (t1 - t0))

```
The inference time I got is:

![image](https://user-images.githubusercontent.com/29330657/88369778-52d18100-cdc3-11ea-8544-114610ecaca8.png)

when I add the sentences below, the time differs much and the new inference time I got is:

```
#NOTE:  This must be the first call in order to work properly!
from nets.colorization.deoldify import device
from nets.colorization.deoldify.device_id import DeviceId
#choices:  CPU, GPU0...GPU7
device.set(device=DeviceId.GPU0)
```

![image](https://user-images.githubusercontent.com/29330657/88369797-5cf37f80-cdc3-11ea-864e-d5c52c5d5b68.png)

As you can see, `os.environ['CUDA_VISIBLE_DEVICES']='0'` is always used, while time is much different, so I wandered if you  noticed this phenomenon and could you give me a hand on this strange issue.

Again, sorry for my unclear explanation!

**Comment by jantic** [comment]
@hiterjoshua I was able to replicate what you're seeing there, but I wouldn't call it an issue really. The whole idea is that the convention is to use the device functionality to use the GPUs. It coordinates the rest of the code in an easy to manage way....provided you stick to the convention.  I was almost tempted to accommodate this scenario by changing the source of the behavior your seeing- this code is called when you import from deoldify.visualize:

```
class _Device:
    def __init__(self):    
        self.set(DeviceId.CPU) 
```

I was about to change this to take into account the user deciding to use this instead:

```
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'
```
Meaning that __init__ would have to do additional logic based on if this 'CUDA_VISIBLE_DEVICES' variable was already set. But then I realized a few things:

1. It will over complicate the code, meaning a greater chance for bugs.  In contrast, what you're seeing now is simply fallback behavior that results from not sticking with the convention- it goes to the CPU.
2. I'd have to take into account the possibility that the user does something this: `os.environ['CUDA_VISIBLE_DEVICES']='0,2,3'`. The problem with trying to accommodate that is two fold:  One, parallel and distributed training/inference aren't supported in the code explicitly so this would be misleading.  Two, the device ids for this reason are oriented to only having one device selectable at a time so doing something like just defaulting to the first device in the list silently would just be misleading. 
3. I'm a big stickler for being disciplined about sticking to convention for understandability reasons, generally. I don't see a point in adding another rabbit hole of "many ways to do one thing"- that just means adding more maintenance and potential bugs to the plate.

Hope that all makes sense!

**Comment by hiterjoshua** [comment]
@jantic At the code level, I totally understand what you talked about, simply, the two code parts should play exactly the same function.
 ```
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'
```

```
#NOTE:  This must be the first call in order to work properly!
from nets.colorization.deoldify import device
from nets.colorization.deoldify.device_id import DeviceId
#choices:  CPU, GPU0...GPU7
device.set(device=DeviceId.GPU0)
```
After I read the codes of this repo, I do agree with you that the two codes were supposed to play the same role and indeed they did in result while differs in inference time! Actually, I didnot see a explanation of the time difference problem convincingly in your previous answer, is that related to some deep mechanism of the nvidia GPU system while we know nothing about it? 
confused......

**Comment by jantic** [comment]
> Actually, I didnot see a explanation of the time difference problem convincingly in your previous answer, is that related to some deep mechanism of the nvidia GPU system while we know nothing about it?

Simply put- my code behind the scenes was changing it to CPU.  That's why it was slower.
